### PR TITLE
[CPU] [DEBUG CAPS] Extension for snippets and other ngraph transformations

### PR DESCRIPTION
**Pull request by imikhale**
### Details:
 - Option to disable specified ngraph transformations.
 - Option to serialize ngraph before and after specified transformations.

### Tickets:
 - 75552

### CODE REVIEW

**Review by tgubanova-lohika — CHANGES_REQUESTED** [review]
Original comment was removed because it violates [A Code of Conduct for Open Source Communities](https://www.contributor-covenant.org/version/2/1/code_of_conduct/)

**Comment by dmitry-gorokhov** [comment]
@EgorDuplensky could you please review?

File: src/plugins/intel_cpu/src/plugin.cpp [hunk 1]
```diff
@@ -171,260 +172,307 @@ Engine::~Engine() {
...
+
+        ConvertToCPUSpecificOpset(nGraphFunc);
+    }
+
+private:
+    const std::shared_ptr<ngraph::Function>& nGraphFunc;
+    using const_node_ptr = const std::shared_ptr<const ngraph::Node>;
+    CPU_DEBUG_CAP_ENABLE(const Config& config);
+
+    void upToCpuSpecificOpSet_common(const bool useLpt, const bool isLegacyApi) {
```
Comment (EgorDuplensky) [inline src/plugins/intel_cpu/src/plugin.cpp:210]: As it was discussed by common transformations we mean ngraph::pass::CommonOptimizations>.
What about the following order:
1) common (CommonOptimizations)
2) cpu essentials or preLpt
3) lpt
4) postLpt
4) snippets

I think there is no need to combine them, we can just call them one after another.

File: src/plugins/intel_cpu/src/utils/ngraph_transformation.hpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        }
+        serializer.run_passes(nGraphFunc);
+        num++;
+    }
+};
+
+}   // namespace intel_cpu
+}   // namespace ov
+
+#  define CPU_DEBUG_CAP_TRANSFORMATION_RETURN_OR_DUMP(_type)                                            \
```
Comment (EgorDuplensky) [inline src/plugins/intel_cpu/src/utils/ngraph_transformation.hpp:78]: I propose to explicitly pass config as a macro parameter

**Review by EgorDuplensky — COMMENTED** [review]
I propose to run benchmark on the changes, just in case we missed some changes in the order of transformations.

File: src/plugins/intel_cpu/src/plugin.cpp (hunk 1 above)
Comment (EgorDuplensky) [reply src/plugins/intel_cpu/src/plugin.cpp:210]: Discussed.
Let's split stages "naturally", with separate steps

Comment (imikhale) [reply src/plugins/intel_cpu/src/plugin.cpp:210]: @EgorDuplensky, is this decision for both disabling and serialization?
that looks good to have the same logic/code for these features.

Comment (EgorDuplensky) [reply src/plugins/intel_cpu/src/plugin.cpp]: leaving this open for final review

File: src/plugins/intel_cpu/src/plugin.cpp [hunk 3]
```diff
@@ -175,268 +176,310 @@ Engine::~Engine() {
...
+        manager.register_pass<ngraph::pass::ConvertNMS1ToNMS5>();
+        manager.register_pass<ngraph::pass::ConvertNMS3ToNMS5>();
+        manager.register_pass<ngraph::pass::ConvertNMS4ToNMS5>();
+        manager.register_pass<ngraph::pass::ConvertNMSToNMSIEInternal>();
+        manager.register_pass<ngraph::pass::ConvertMulticlassNmsToMulticlassNmsIE>();
+        manager.register_pass<ngraph::pass::ConvertMatrixNmsToMatrixNmsIE>();
+        manager.register_pass<ngraph::pass::TransposeMatMul>();
+        manager.register_pass<ngraph::pass::ConstantFolding>();
+
+        if (useLpt) {
```
Comment (EgorDuplensky) [inline src/plugins/intel_cpu/src/plugin.cpp:283]: It seems we are not completely disabling lpt
I think we should set useLpt to false when disable Lpt transformation
For example using some additional marco, like
CPU_DEBUG_CAP_DISABLE_IF(useLpt, condition) or something similar.

File: src/plugins/intel_cpu/src/config.h [hunk 4]
```diff
@@ -47,6 +50,146 @@ struct Config {
     std::map<std::string, std::string> _config;
 
 #ifdef CPU_DEBUG_CAPS
+
+private:
```
[the remark this answers is not in the document]
Comment (imikhale) [reply src/plugins/intel_cpu/src/config.h:54]: well, let's do it later

File: src/plugins/intel_cpu/src/plugin.cpp (hunk 3 above)
Comment (imikhale) [reply src/plugins/intel_cpu/src/plugin.cpp:285]: missed it, thank you, redesigned

File: src/plugins/intel_cpu/src/plugin.cpp [hunk 5]
```diff
@@ -175,268 +176,310 @@ Engine::~Engine() {
...
-    const bool useLpt =
-            _enableLPT &&
-        ngraph::pass::low_precision::LowPrecision::isFunctionQuantized(nGraphFunc);
-    auto defaultPrecisions = useLpt ? ngraph::pass::low_precision::precision_set::int8_support : std::vector<ov::element::Type>{};
-    bool hasINT16orINT32Levels = false;
-    if (useLpt) {
+class Transformations {
+public:
+    Transformations(const std::shared_ptr<ngraph::Function>& ngraphFunction
+                    CPU_DEBUG_CAP_ENABLE(, const Config& config))
```
[the remark this answers is not in the document]
Comment (imikhale) [reply src/plugins/intel_cpu/src/plugin.cpp]: yep, I don't like such macro tricks too, removed assuming that this tiny performance penalty is acceptable

**Comment by dmitry-gorokhov** [comment]
Implemented in https://github.com/openvinotoolkit/openvino/pull/14223

**Pull request closed without merging**